
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Vishal Koparde, Ph.D.">
      
      
      
        <link rel="prev" href="../flowchart/">
      
      
        <link rel="next" href="../references/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.1.6">
    
    
      
        <title>Tutorial - CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.ded33207.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#charlie" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)" class="md-header__button md-logo" aria-label="CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/CCBR/CHARLIE" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    CCBR/CHARLIE
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)" class="md-nav__button md-logo" aria-label="CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    CHARLIE (Circrnas in Host And viRuses anaLysis pIpEline)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CCBR/CHARLIE" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    CCBR/CHARLIE
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../flowchart/" class="md-nav__link">
        Flowchart
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Tutorial
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial" class="md-nav__link">
    Tutorial
  </a>
  
    <nav class="md-nav" aria-label="Tutorial">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    Prerequisites
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#location" class="md-nav__link">
    Location
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#init" class="md-nav__link">
    Init
  </a>
  
    <nav class="md-nav" aria-label="Init">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#configyaml" class="md-nav__link">
    config.yaml
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#samplestsv" class="md-nav__link">
    samples.tsv
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dryrun" class="md-nav__link">
    Dryrun
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run" class="md-nav__link">
    Run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-tracking" class="md-nav__link">
    Error tracking
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#expected-output" class="md-nav__link">
    Expected output:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../references/" class="md-nav__link">
        References
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../versions/" class="md-nav__link">
        Versions
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial" class="md-nav__link">
    Tutorial
  </a>
  
    <nav class="md-nav" aria-label="Tutorial">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    Prerequisites
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#location" class="md-nav__link">
    Location
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#init" class="md-nav__link">
    Init
  </a>
  
    <nav class="md-nav" aria-label="Init">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#configyaml" class="md-nav__link">
    config.yaml
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#samplestsv" class="md-nav__link">
    samples.tsv
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dryrun" class="md-nav__link">
    Dryrun
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run" class="md-nav__link">
    Run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-tracking" class="md-nav__link">
    Error tracking
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#expected-output" class="md-nav__link">
    Expected output:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="charlie">CHARLIE</h1>
<p><img alt="img" src="https://img.shields.io/github/issues/kopardev/circRNA?style=for-the-badge" /><img alt="img" src="https://img.shields.io/github/forks/kopardev/circRNA?style=for-the-badge" /><img alt="img" src="https://img.shields.io/github/stars/kopardev/circRNA?style=for-the-badge" /><img alt="img" src="https://img.shields.io/github/license/kopardev/circRNA?style=for-the-badge" /></p>
<h3 id="tutorial">Tutorial</h3>
<h4 id="prerequisites">Prerequisites</h4>
<ul>
<li>
<p><a href="https://hpc.nih.gov/">Biowulf</a> account: Biowulf account can be requested <a href="https://hpc.nih.gov/docs/accounts.html">here</a>.</p>
</li>
<li>
<p>Membership to Ziegelbauer user group on Biowulf. You can check this by typing the following command:</p>
</li>
</ul>
<p><code>bash
  % groups</code></p>
<p>output:</p>
<p><code>bash
  CCBR kopardevn Ziegelbauer_lab</code></p>
<p>If <code>Ziegelbauer_lab</code> is not listed then you can email a request to be added to the groups <a href="mailto:staff@hpc.nih.gov">here</a></p>
<h4 id="location">Location</h4>
<p>Different versions of circRNA DAQ pipeline have been parked at <code>/data/Ziegelbauer_lab/Pipelines/circRNA</code></p>
<pre><code class="language-bash">% ls /data/Ziegelbauer_lab/Pipelines/circRNA
</code></pre>
<p>output:</p>
<pre><code class="language-bash">v0.1.0
v0.10.0
v0.10.0-dev
v0.2.1
v0.3.3
v0.4.2
v0.5.2
v0.6.5
v0.7.0
v0.8
v0.9.0
</code></pre>
<p>The exacts versions listed here may changed as newer versions are added. Also, the <code>dev</code> version is pointing to the most recent untagged version of the pipeline (use at own risk!)</p>
<h4 id="init">Init</h4>
<p>To get help about the pipeline you can run:</p>
<pre><code class="language-bash">% bash /data/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev/charlie
</code></pre>
<p>output:</p>
<pre><code class="language-bash">
##########################################################################################

Welcome to charlie(v0.10.0-dev)
 _______  __   __  _______  ______    ___      ___   _______
|       ||  | |  ||   _   ||    _ |  |   |    |   | |       |
|       ||  |_|  ||  |_|  ||   | ||  |   |    |   | |    ___|
|       ||       ||       ||   |_||_ |   |    |   | |   |___
|      _||       ||       ||    __  ||   |___ |   | |    ___|
|     |_ |   _   ||   _   ||   |  | ||       ||   | |   |___
|_______||__| |__||__| |__||___|  |_||_______||___| |_______|

C_ircrnas in H_ost A_nd vi_R_uses ana_L_ysis p_I_p_E_line

##########################################################################################

This pipeline was built by CCBR (https://bioinformatics.ccr.cancer.gov/ccbr)
Please contact Vishal Koparde for comments/questions (vishal.koparde@nih.gov)

##########################################################################################

CHARLIE can be used to DAQ(Detect/Annotate/Quantify) circRNAs in hosts and viruses.

Here is the list of hosts and viruses that are currently supported:

HOSTS:
  * hg38          [Human]
  * mm39          [Mouse]

ADDITIVES:
  * ERCC          [External RNA Control Consortium sequences]
  * BAC16Insert   [insert from rKSHV.219-derived BAC clone of the full-length KSHV genome]

VIRUSES:
  * NC_007605.1   [Human gammaherpesvirus 4 (Epstein-Barr virus)]
  * NC_006273.2   [Human betaherpesvirus 5 (Cytomegalovirus )]
  * NC_001664.4   [Human betaherpesvirus 6A (HHV-6A)]
  * NC_000898.1   [Human betaherpesvirus 6B (HHV-6B)]
  * NC_001716.2   [Human betaherpesvirus 7 (HHV-7)]
  * NC_009333.1   [Human gammaherpesvirus 8 (KSHV)]
  * NC_045512.2   [Severe acute respiratory syndrome(SARS)-related coronavirus]
  * MN485971.1    [HIV from Belgium]
  * NC_001806.2   [Human alphaherpesvirus 1 (Herpes simplex virus type 1)](strain 17) (HSV-1)]
  * KT899744.1    [HSV-1 strain KOS]
  * MH636806.1    [MHV68 (Murine herpesvirus 68 strain WUMS)]

##########################################################################################

USAGE:
  bash /data/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev/charlie -w/--workdir=&lt;WORKDIR&gt; -m/--runmode=&lt;RUNMODE&gt;

Required Arguments:
1.  WORKDIR     : [Type: String]: Absolute or relative path to the output folder with write permissions.

2.  RUNMODE     : [Type: String] Valid options:
    * init      : initialize workdir
    * dryrun    : dry run snakemake to generate DAG
    * run       : run with slurm
    * runlocal  : run without submitting to sbatch
    ADVANCED RUNMODES (use with caution!!)
    * unlock    : unlock WORKDIR if locked by snakemake NEVER UNLOCK WORKDIR WHERE PIPELINE IS CURRENTLY RUNNING!
    * reconfig  : recreate config file in WORKDIR (debugging option) EDITS TO config.yaml WILL BE LOST!
    * reset     : DELETE workdir dir and re-init it (debugging option) EDITS TO ALL FILES IN WORKDIR WILL BE LOST!
    * printbinds: print singularity binds (paths)
    * local     : same as runlocal

Optional Arguments:

--host|-g       : supply host at command line. hg38 or mm39.                                            (--runmode=init only)
--additives|-a  : supply comma-separated list of additives at command line. ERCC or BAC16Insert or both (--runmode=init only)
--viruses|-v    : supply comma-separated list of viruses at command line                                (--runmode=init only)
--manifest|-s   : absolute path to samples.tsv. This will be copied to output folder                    (--runmode=init only)
--changegrp|-z  : change group to &quot;Ziegelbauer_lab&quot; before running anything. Biowulf-only. Useful for correctly setting permissions.
--help|-h       : print this help


Example commands:
  bash /data/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev/charlie -w=/my/ouput/folder -m=init
  bash /data/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev/charlie -w=/my/ouput/folder -m=dryrun
  bash /data/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev/charlie -w=/my/ouput/folder -m=run

##########################################################################################

VersionInfo:
  python          : 3.7
  snakemake       : 7.19.1
  pipeline_home   : /vf/users/Ziegelbauer_lab/Pipelines/circRNA/v0.10.0-dev
  git commit/tag  : b2cf2f089788651041b16bf4378c2c5172c13cb2    v0.10.0-2-gb2cf2f0

##########################################################################################
</code></pre>
<blockquote>
<p>NOTE:
You can replace <code>v0.10.0</code> in the above command with the latest version to use a newer version. <code>run_circrna_daq.sh</code> was called <code>test.sh</code> in versions older than <code>v0.4.0</code>.</p>
</blockquote>
<p>To initial the working directory run:</p>
<pre><code class="language-bash">% bash &lt;path to charlie&gt; -w=&lt;path to output dir&gt; -m=init
</code></pre>
<p>This assumes that <code>&lt;path to output dir&gt;</code> does not exist before running the above command and is at a location where write permissions are available.</p>
<p>The above command creates <code>&lt;path to output dir&gt;</code> folder and creates 2 subfolders <code>logs</code> and <code>stats</code> inside that folder along with <code>config.yaml</code> and <code>samples.tsv</code> files.</p>
<pre><code class="language-bash">% tree &lt;path to output dir&gt;
</code></pre>
<h5 id="configyaml">config.yaml</h5>
<p>This file is used to fine tune the execution of the pipeline by setting:</p>
<ul>
<li>sample sheet location ... aka <code>samples.tsv</code></li>
<li>which circRNA finding tools to use by editing these:<ul>
<li>run_clear: True</li>
<li>run_dcc: True</li>
<li>run_mapsplice: False</li>
<li>run_circRNAFinder: True</li>
<li>run_nclscan: False</li>
<li>run_findcirc: False</li>
</ul>
</li>
<li>describes the location of other resources/indexes/tools etc. Generally, these do NOT need to be changed.</li>
</ul>
<h5 id="samplestsv">samples.tsv</h5>
<p>Tab delimited definition of sample sheet. The header is fixed and each row represents a sample. It has 3 columns:</p>
<ol>
<li>sampleName = Name of the sample. This has to be unique.</li>
<li>path_to_R1_fastq = absolute path to the read1 fastq.gz file.</li>
<li>path_to_R2_fastq = absolute path to the read2 fastq.gz file. If the sample was sequenced in single-end mode, then leave this blank.</li>
</ol>
<p>The <code>.tests/dummy_fastqs</code> folder in the repo has test dataset:</p>
<pre><code class="language-bash">% tree .tests/dummy_fastqs
.tests/dummy_fastqs
├── GI1_N.R1.fastq.gz
├── GI1_N.R2.fastq.gz
└── GI1_T.R1.fastq.gz
</code></pre>
<p><code>GI1_N</code> is a PE sample while <code>GI1_T</code> is a SE sample.</p>
<h4 id="dryrun">Dryrun</h4>
<p>Once the <code>samples.tsv</code> file has been edited appropriately to include the desired samples, it is a good idea to <strong>dryrun</strong> the pipeline to ensure that everything will work as desired. Dryrun can be run as follows:</p>
<pre><code class="language-bash">bash &lt;path to charlie&gt; -w=&lt;path to output dir&gt; -m=dryrun
</code></pre>
<p>This will create the reference fasta and gtf file based on the selections made in the <code>config.yaml</code>. Hence, can take a few minutes to run.</p>
<h4 id="run">Run</h4>
<p>Upon verifying that dryrun is successful. You can then submit the job to the cluster using the following command:</p>
<pre><code class="language-bash">bash &lt;path to charlie&gt; -w=&lt;path to output dir&gt; -m=run
</code></pre>
<p>which will produce something like this:</p>
<pre><code>... ... skipping ~1000 lines
...
...
Job stats:
job                                              count    min threads    max threads
---------------------------------------------  -------  ----------
all                                                  1              1              1
annotate_clear_output                                2              1              1
circExplorer                                         2              2              2
circExplorer_bwa                                     2              2              2
circrnafinder                                        2              1              1
ciri                                                 2             56             56
clear                                                2              2              2
create_bowtie2_index                                 1              1              1
create_bwa_index                                     1              1              1
create_circExplorer_BSJ_bam                          2              4              4
create_circExplorer_linear_spliced_bams              2             56             56
create_circExplorer_merged_found_counts_table        2              1              1
create_hq_bams                                       2              1              1
create_index                                         1             56             56
create_master_counts_file                            1              1              1
cutadapt                                             2             56             56
dcc                                                  2              4              4
dcc_create_samplesheets                              2              1              1
estimate_duplication                                 2              1              1
fastqc                                               2              4              4
find_circ                                            2             56             56
find_circ_align                                      2             56             56
merge_SJ_tabs                                        1              2              2
merge_alignment_stats                                1              1              1
merge_genecounts                                     1              1              1
merge_per_sample                                     2              1              1
star1p                                               2             56             56
star2p                                               2             56             56
star_circrnafinder                                   2             56             56
total                                               52              1             56

Reasons:
    (check individual jobs above for details)
    input files updated by another job:
        alignment_stats, all, annotate_clear_output, circExplorer, circExplorer_bwa, circrnafinder, ciri, clear, create_circExplorer_BSJ_bam, create_circExplorer_linear_spliced_bams, create_circExplorer_merged_found_counts_table, create_hq_bams, create_master_counts_file, dcc, dcc_create_samplesheets, estimate_duplication, fastqc, find_circ, find_circ_align, merge_SJ_tabs, merge_alignment_stats, merge_genecounts, merge_per_sample, star1p, star2p, star_circrnafinder
    missing output files:
        alignment_stats, annotate_clear_output, circExplorer, circExplorer_bwa, circrnafinder, ciri, clear, create_bowtie2_index, create_bwa_index, create_circExplorer_BSJ_bam, create_circExplorer_linear_spliced_bams, create_circExplorer_merged_found_counts_table, create_hq_bams, create_index, create_master_counts_file, cutadapt, dcc, dcc_create_samplesheets, estimate_duplication, fastqc, find_circ, find_circ_align, merge_SJ_tabs, merge_alignment_stats, merge_genecounts, merge_per_sample, star1p, star2p, star_circrnafinder

This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
Running...
14743440
</code></pre>
<p>In this example, <code>14743440</code> is the jobid returned by the slurm job scheduler on biowulf. This means that the job was successfully submitted, it will spawn off other subjobs which in-turn will be run and outputs will be moved to the <code>results</code> folder created inside the working directory supplied at command line. You can check the status of your queue of jobs in biowulf running:</p>
<pre><code class="language-bash">% squeue -u `whoami`
</code></pre>
<p>output:</p>
<pre><code class="language-bash">             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           14743440  ccr,norm  circRNA kopardev PD       0:00      1 (None)
</code></pre>
<p><code>ST</code> in the above results stands for Status and <code>PD</code> means Pending. The status will change from pending(<code>PD</code>) to running(<code>R</code>) to completed as jobs are run on the cluster.</p>
<p>Next, just sit tight until the pipeline finishes. You can keep monitoring the queue as shown above. If there are no jobs running on biowulf, then your pipeline has finished (or errored out!)</p>
<p>Once completed the output should something like this:</p>
<pre><code class="language-bash">% tree &lt;path to output dir&gt;
</code></pre>
<p>output:</p>
<pre><code class="language-bash">.
├── cluster.json
├── config.yaml
├── dryrun.231222103505.log
├── fastqs
│   ├── GI1_N.R1.fastq.gz -&gt; /data/CCBR_Pipeliner/testdata/circRNA/human/GI1_N_ss.R1.fastq.gz
│   ├── GI1_N.R2.fastq.gz -&gt; /data/CCBR_Pipeliner/testdata/circRNA/human/GI1_N_ss.R2.fastq.gz
│   └── GI1_T.R1.fastq.gz -&gt; /data/CCBR_Pipeliner/testdata/circRNA/human/GI1_T_ss.R1.fastq.gz
├── jobby
├── jobinfo.short
├── jobinfo.txt
├── logs
│   ... log files ... 
│   ... skipping ...
├── nclscan.config
├── qc
│   ├── fastqc
│   │   ├── GI1_N.R1_fastqc.html
│   │   ├── GI1_N.R1_fastqc.zip
│   │   ├── GI1_N.R1.trim_fastqc.html
│   │   ├── GI1_N.R1.trim_fastqc.zip
│   │   ├── GI1_N.R2_fastqc.html
│   │   ├── GI1_N.R2_fastqc.zip
│   │   ├── GI1_N.R2.trim_fastqc.html
│   │   ├── GI1_N.R2.trim_fastqc.zip
│   │   ├── GI1_T.R1_fastqc.html
│   │   ├── GI1_T.R1_fastqc.zip
│   │   ├── GI1_T.R1.trim_fastqc.html
│   │   └── GI1_T.R1.trim_fastqc.zip
│   └── picard_MarkDuplicates
│       ├── GI1_N.MarkDuplicates.metrics.txt
│       └── GI1_T.MarkDuplicates.metrics.txt
├── ref
│   ├── bwa_index.log
│   ├── gene_id_2_gene_name.tsv
│   ├── NCLscan_index
│   │   ├── AllRef.fa
│   │   ├── AllRef.fa.amb
│   │   ├── AllRef.fa.ann
│   │   ├── AllRef.fa.bwt
│   │   ├── AllRef.fa.pac
│   │   ├── AllRef.fa.sa
│   │   ├── AllRef.ndx
│   │   └── RepChrM.fa
│   ├── ref.1.bt2
│   ├── ref.2.bt2
│   ├── ref.3.bt2
│   ├── ref.4.bt2
│   ├── ref.amb
│   ├── ref.ann
│   ├── ref.bwt
│   ├── ref.dummy.fa
│   ├── ref.fa
│   ├── ref.fa.byo_index
│   ├── ref.fa.fai
│   ├── ref.fa.regions
│   ├── ref.fa.regions.host
│   ├── ref.fa.regions.viruses
│   ├── ref.fa.sizes
│   ├── ref.fixed.gtf
│   ├── ref.genes.genepred
│   ├── ref.genes.genepred_w_geneid
│   ├── ref.gtf
│   ├── ref.pac
│   ├── ref.rev.1.bt2
│   ├── ref.rev.2.bt2
│   ├── ref.sa
│   ├── ref.transcripts.fa
│   ├── separate_fastas
│   │   ├── chr10.fa
│   │   ├── chr11.fa
│   │   ├── chr12.fa
│   │   ├── chr13.fa
│   │   ├── chr14.fa
│   │   ├── chr15.fa
│   │   ├── chr16.fa
│   │   ├── chr17.fa
│   │   ├── chr18.fa
│   │   ├── chr19.fa
│   │   ├── chr1.fa
│   │   ├── chr20.fa
│   │   ├── chr21.fa
│   │   ├── chr22.fa
│   │   ├── chr2.fa
│   │   ├── chr3.fa
│   │   ├── chr45S.fa
│   │   ├── chr4.fa
│   │   ├── chr5.fa
│   │   ├── chr5S.fa
│   │   ├── chr6.fa
│   │   ├── chr7.fa
│   │   ├── chr8.fa
│   │   ├── chr9.fa
│   │   ├── chrM.fa
│   │   ├── chrX.fa
│   │   ├── chrY.fa
│   │   ├── ERCC_00002.fa
│   │   ├── ERCC_00003.fa
│   │   ├── ERCC_00004.fa
│   │   ├── ERCC_00009.fa
│   │   ├── ERCC_00012.fa
│   │   ├── ERCC_00013.fa
│   │   ├── ERCC_00014.fa
│   │   ├── ERCC_00016.fa
│   │   ├── ERCC_00017.fa
│   │   ├── ERCC_00019.fa
│   │   ├── ERCC_00022.fa
│   │   ├── ERCC_00024.fa
│   │   ├── ERCC_00025.fa
│   │   ├── ERCC_00028.fa
│   │   ├── ERCC_00031.fa
│   │   ├── ERCC_00033.fa
│   │   ├── ERCC_00034.fa
│   │   ├── ERCC_00035.fa
│   │   ├── ERCC_00039.fa
│   │   ├── ERCC_00040.fa
│   │   ├── ERCC_00041.fa
│   │   ├── ERCC_00042.fa
│   │   ├── ERCC_00043.fa
│   │   ├── ERCC_00044.fa
│   │   ├── ERCC_00046.fa
│   │   ├── ERCC_00048.fa
│   │   ├── ERCC_00051.fa
│   │   ├── ERCC_00053.fa
│   │   ├── ERCC_00054.fa
│   │   ├── ERCC_00057.fa
│   │   ├── ERCC_00058.fa
│   │   ├── ERCC_00059.fa
│   │   ├── ERCC_00060.fa
│   │   ├── ERCC_00061.fa
│   │   ├── ERCC_00062.fa
│   │   ├── ERCC_00067.fa
│   │   ├── ERCC_00069.fa
│   │   ├── ERCC_00071.fa
│   │   ├── ERCC_00073.fa
│   │   ├── ERCC_00074.fa
│   │   ├── ERCC_00075.fa
│   │   ├── ERCC_00076.fa
│   │   ├── ERCC_00077.fa
│   │   ├── ERCC_00078.fa
│   │   ├── ERCC_00079.fa
│   │   ├── ERCC_00081.fa
│   │   ├── ERCC_00083.fa
│   │   ├── ERCC_00084.fa
│   │   ├── ERCC_00085.fa
│   │   ├── ERCC_00086.fa
│   │   ├── ERCC_00092.fa
│   │   ├── ERCC_00095.fa
│   │   ├── ERCC_00096.fa
│   │   ├── ERCC_00097.fa
│   │   ├── ERCC_00098.fa
│   │   ├── ERCC_00099.fa
│   │   ├── ERCC_00104.fa
│   │   ├── ERCC_00108.fa
│   │   ├── ERCC_00109.fa
│   │   ├── ERCC_00111.fa
│   │   ├── ERCC_00112.fa
│   │   ├── ERCC_00113.fa
│   │   ├── ERCC_00116.fa
│   │   ├── ERCC_00117.fa
│   │   ├── ERCC_00120.fa
│   │   ├── ERCC_00123.fa
│   │   ├── ERCC_00126.fa
│   │   ├── ERCC_00130.fa
│   │   ├── ERCC_00131.fa
│   │   ├── ERCC_00134.fa
│   │   ├── ERCC_00136.fa
│   │   ├── ERCC_00137.fa
│   │   ├── ERCC_00138.fa
│   │   ├── ERCC_00142.fa
│   │   ├── ERCC_00143.fa
│   │   ├── ERCC_00144.fa
│   │   ├── ERCC_00145.fa
│   │   ├── ERCC_00147.fa
│   │   ├── ERCC_00148.fa
│   │   ├── ERCC_00150.fa
│   │   ├── ERCC_00154.fa
│   │   ├── ERCC_00156.fa
│   │   ├── ERCC_00157.fa
│   │   ├── ERCC_00158.fa
│   │   ├── ERCC_00160.fa
│   │   ├── ERCC_00162.fa
│   │   ├── ERCC_00163.fa
│   │   ├── ERCC_00164.fa
│   │   ├── ERCC_00165.fa
│   │   ├── ERCC_00168.fa
│   │   ├── ERCC_00170.fa
│   │   ├── ERCC_00171.fa
│   │   ├── GL000008.2.fa
│   │   ├── GL000009.2.fa
│   │   ├── GL000194.1.fa
│   │   ├── GL000195.1.fa
│   │   ├── GL000205.2.fa
│   │   ├── GL000208.1.fa
│   │   ├── GL000213.1.fa
│   │   ├── GL000214.1.fa
│   │   ├── GL000216.2.fa
│   │   ├── GL000218.1.fa
│   │   ├── GL000219.1.fa
│   │   ├── GL000220.1.fa
│   │   ├── GL000221.1.fa
│   │   ├── GL000224.1.fa
│   │   ├── GL000225.1.fa
│   │   ├── GL000226.1.fa
│   │   ├── KI270302.1.fa
│   │   ├── KI270303.1.fa
│   │   ├── KI270304.1.fa
│   │   ├── KI270305.1.fa
│   │   ├── KI270310.1.fa
│   │   ├── KI270311.1.fa
│   │   ├── KI270312.1.fa
│   │   ├── KI270315.1.fa
│   │   ├── KI270316.1.fa
│   │   ├── KI270317.1.fa
│   │   ├── KI270320.1.fa
│   │   ├── KI270322.1.fa
│   │   ├── KI270329.1.fa
│   │   ├── KI270330.1.fa
│   │   ├── KI270333.1.fa
│   │   ├── KI270334.1.fa
│   │   ├── KI270335.1.fa
│   │   ├── KI270336.1.fa
│   │   ├── KI270337.1.fa
│   │   ├── KI270338.1.fa
│   │   ├── KI270340.1.fa
│   │   ├── KI270362.1.fa
│   │   ├── KI270363.1.fa
│   │   ├── KI270364.1.fa
│   │   ├── KI270366.1.fa
│   │   ├── KI270371.1.fa
│   │   ├── KI270372.1.fa
│   │   ├── KI270373.1.fa
│   │   ├── KI270374.1.fa
│   │   ├── KI270375.1.fa
│   │   ├── KI270376.1.fa
│   │   ├── KI270378.1.fa
│   │   ├── KI270379.1.fa
│   │   ├── KI270381.1.fa
│   │   ├── KI270382.1.fa
│   │   ├── KI270383.1.fa
│   │   ├── KI270384.1.fa
│   │   ├── KI270385.1.fa
│   │   ├── KI270386.1.fa
│   │   ├── KI270387.1.fa
│   │   ├── KI270388.1.fa
│   │   ├── KI270389.1.fa
│   │   ├── KI270390.1.fa
│   │   ├── KI270391.1.fa
│   │   ├── KI270392.1.fa
│   │   ├── KI270393.1.fa
│   │   ├── KI270394.1.fa
│   │   ├── KI270395.1.fa
│   │   ├── KI270396.1.fa
│   │   ├── KI270411.1.fa
│   │   ├── KI270412.1.fa
│   │   ├── KI270414.1.fa
│   │   ├── KI270417.1.fa
│   │   ├── KI270418.1.fa
│   │   ├── KI270419.1.fa
│   │   ├── KI270420.1.fa
│   │   ├── KI270422.1.fa
│   │   ├── KI270423.1.fa
│   │   ├── KI270424.1.fa
│   │   ├── KI270425.1.fa
│   │   ├── KI270429.1.fa
│   │   ├── KI270435.1.fa
│   │   ├── KI270438.1.fa
│   │   ├── KI270442.1.fa
│   │   ├── KI270448.1.fa
│   │   ├── KI270465.1.fa
│   │   ├── KI270466.1.fa
│   │   ├── KI270467.1.fa
│   │   ├── KI270468.1.fa
│   │   ├── KI270507.1.fa
│   │   ├── KI270508.1.fa
│   │   ├── KI270509.1.fa
│   │   ├── KI270510.1.fa
│   │   ├── KI270511.1.fa
│   │   ├── KI270512.1.fa
│   │   ├── KI270515.1.fa
│   │   ├── KI270516.1.fa
│   │   ├── KI270517.1.fa
│   │   ├── KI270518.1.fa
│   │   ├── KI270519.1.fa
│   │   ├── KI270521.1.fa
│   │   ├── KI270522.1.fa
│   │   ├── KI270528.1.fa
│   │   ├── KI270529.1.fa
│   │   ├── KI270530.1.fa
│   │   ├── KI270538.1.fa
│   │   ├── KI270539.1.fa
│   │   ├── KI270544.1.fa
│   │   ├── KI270548.1.fa
│   │   ├── KI270579.1.fa
│   │   ├── KI270580.1.fa
│   │   ├── KI270581.1.fa
│   │   ├── KI270582.1.fa
│   │   ├── KI270583.1.fa
│   │   ├── KI270584.1.fa
│   │   ├── KI270587.1.fa
│   │   ├── KI270588.1.fa
│   │   ├── KI270589.1.fa
│   │   ├── KI270590.1.fa
│   │   ├── KI270591.1.fa
│   │   ├── KI270593.1.fa
│   │   ├── KI270706.1.fa
│   │   ├── KI270707.1.fa
│   │   ├── KI270708.1.fa
│   │   ├── KI270709.1.fa
│   │   ├── KI270710.1.fa
│   │   ├── KI270711.1.fa
│   │   ├── KI270712.1.fa
│   │   ├── KI270713.1.fa
│   │   ├── KI270714.1.fa
│   │   ├── KI270715.1.fa
│   │   ├── KI270716.1.fa
│   │   ├── KI270717.1.fa
│   │   ├── KI270718.1.fa
│   │   ├── KI270719.1.fa
│   │   ├── KI270720.1.fa
│   │   ├── KI270721.1.fa
│   │   ├── KI270722.1.fa
│   │   ├── KI270723.1.fa
│   │   ├── KI270724.1.fa
│   │   ├── KI270725.1.fa
│   │   ├── KI270726.1.fa
│   │   ├── KI270727.1.fa
│   │   ├── KI270728.1.fa
│   │   ├── KI270729.1.fa
│   │   ├── KI270730.1.fa
│   │   ├── KI270731.1.fa
│   │   ├── KI270732.1.fa
│   │   ├── KI270733.1.fa
│   │   ├── KI270734.1.fa
│   │   ├── KI270735.1.fa
│   │   ├── KI270736.1.fa
│   │   ├── KI270737.1.fa
│   │   ├── KI270738.1.fa
│   │   ├── KI270739.1.fa
│   │   ├── KI270740.1.fa
│   │   ├── KI270741.1.fa
│   │   ├── KI270742.1.fa
│   │   ├── KI270743.1.fa
│   │   ├── KI270744.1.fa
│   │   ├── KI270745.1.fa
│   │   ├── KI270746.1.fa
│   │   ├── KI270747.1.fa
│   │   ├── KI270748.1.fa
│   │   ├── KI270749.1.fa
│   │   ├── KI270750.1.fa
│   │   ├── KI270751.1.fa
│   │   ├── KI270752.1.fa
│   │   ├── KI270753.1.fa
│   │   ├── KI270754.1.fa
│   │   ├── KI270755.1.fa
│   │   ├── KI270756.1.fa
│   │   ├── KI270757.1.fa
│   │   ├── NC_009333.1.fa
│   │   └── separate_fastas.lst
│   └── STAR_no_GTF
│       ├── chrLength.txt
│       ├── chrNameLength.txt
│       ├── chrName.txt
│       ├── chrStart.txt
│       ├── Genome
│       ├── genomeParameters.txt
│       ├── Log.out
│       ├── SA
│       └── SAindex
├── results
│   ├── alignmentstats.txt
│   ├── circRNA_master_counts.tsv.gz
│   ├── GI1_N
│   │   ├── alignmentstats.txt
│   │   ├── circExplorer
│   │   │   ├── GI1_N.back_spliced_junction.bed
│   │   │   ├── GI1_N.back_spliced_junction.strand_fixed.bed
│   │   │   ├── GI1_N.bam
│   │   │   ├── GI1_N.bam.bai
│   │   │   ├── GI1_N.BSJ.bam
│   │   │   ├── GI1_N.BSJ.bam.csi
│   │   │   ├── GI1_N.BSJ.bed.gz
│   │   │   ├── GI1_N.BSJ.bw
│   │   │   ├── GI1_N.BSJ.foundcounts.tsv
│   │   │   ├── GI1_N.BSJ.minus.bam
│   │   │   ├── GI1_N.BSJ.minus.bam.csi
│   │   │   ├── GI1_N.BSJ.minus.bw
│   │   │   ├── GI1_N.BSJ.plus.bam
│   │   │   ├── GI1_N.BSJ.plus.bam.csi
│   │   │   ├── GI1_N.BSJ.plus.bw
│   │   │   ├── GI1_N.circExplorer.annotation_counts.tsv
│   │   │   ├── GI1_N.circExplorer.counts_table.tsv
│   │   │   ├── GI1_N_circexplorer_parse.log
│   │   │   ├── GI1_N.circRNA_known.filter1.txt
│   │   │   ├── GI1_N.circularRNA_known.txt
│   │   │   ├── GI1_N.hg38.BSJ.bam
│   │   │   ├── GI1_N.hg38.BSJ.bam.csi
│   │   │   ├── GI1_N.hg38.BSJ.bw
│   │   │   ├── GI1_N.linear.bam
│   │   │   ├── GI1_N.linear.bam.csi
│   │   │   ├── GI1_N.linear_BSJ.bam
│   │   │   ├── GI1_N.linear_BSJ.bam.csi
│   │   │   ├── GI1_N.linear_BSJ.bw
│   │   │   ├── GI1_N.linear_BSJ.hg38.bam
│   │   │   ├── GI1_N.linear_BSJ.hg38.bw
│   │   │   ├── GI1_N.linear_BSJ.NC_009333.1.bam
│   │   │   ├── GI1_N.linear_BSJ.NC_009333.1.bw
│   │   │   ├── GI1_N.linear.BSJ.readids.gz
│   │   │   ├── GI1_N.linear.bw
│   │   │   ├── GI1_N.linear.hg38.bam
│   │   │   ├── GI1_N.linear.hg38.bw
│   │   │   ├── GI1_N.linear.NC_009333.1.bam
│   │   │   ├── GI1_N.linear.NC_009333.1.bw
│   │   │   ├── GI1_N.linear_spliced.counts.tsv
│   │   │   ├── GI1_N.NC_009333.1.BSJ.bam
│   │   │   ├── GI1_N.NC_009333.1.BSJ.bam.csi
│   │   │   ├── GI1_N.NC_009333.1.BSJ.bw
│   │   │   ├── GI1_N.readcounts.tsv
│   │   │   ├── GI1_N.rid2jid.tsv.gz
│   │   │   ├── GI1_N.spliced.bam
│   │   │   ├── GI1_N.spliced.bam.csi
│   │   │   ├── GI1_N.spliced_BSJ.bam
│   │   │   ├── GI1_N.spliced_BSJ.bam.csi
│   │   │   ├── GI1_N.spliced_BSJ.bw
│   │   │   ├── GI1_N.spliced_BSJ.hg38.bam
│   │   │   ├── GI1_N.spliced_BSJ.hg38.bw
│   │   │   ├── GI1_N.spliced_BSJ.NC_009333.1.bam
│   │   │   ├── GI1_N.spliced_BSJ.NC_009333.1.bw
│   │   │   ├── GI1_N.spliced.BSJ.readids.gz
│   │   │   ├── GI1_N.spliced.bw
│   │   │   ├── GI1_N.spliced.hg38.bam
│   │   │   ├── GI1_N.spliced.hg38.bw
│   │   │   ├── GI1_N.spliced.NC_009333.1.bam
│   │   │   ├── GI1_N.spliced.NC_009333.1.bw
│   │   │   ├── low_conf_GI1_N.circRNA_known.filter1.txt
│   │   │   └── low_conf_GI1_N.circRNA_known.txt
│   │   ├── circExplorer_BWA
│   │   │   ├── GI1_N.back_spliced_junction.bed
│   │   │   ├── GI1_N.back_spliced_junction.strand_fixed.bed
│   │   │   ├── GI1_N.circExplorer_bwa.annotation_counts.tsv
│   │   │   ├── GI1_N_circexplorer_parse.log
│   │   │   ├── GI1_N.circRNA_known.filter1.txt
│   │   │   ├── GI1_N.circularRNA_known.txt
│   │   │   ├── low_conf_GI1_N.circRNA_known.filter1.txt
│   │   │   └── low_conf_GI1_N.circRNA_known.txt
│   │   ├── circRNA_finder
│   │   │   ├── GI1_N.Chimeric.out.sorted.bam
│   │   │   ├── GI1_N.Chimeric.out.sorted.bam.bai
│   │   │   ├── GI1_N.circRNA_finder.counts_table.tsv.filtered
│   │   │   ├── GI1_N.filteredJunctions.bed
│   │   │   ├── GI1_N.s_filteredJunctions.bed
│   │   │   └── GI1_N.s_filteredJunctions_fw.bed
│   │   ├── ciri
│   │   │   ├── CIRIerror.log
│   │   │   ├── GI1_N.bwa.log
│   │   │   ├── GI1_N.ciri.bam
│   │   │   ├── GI1_N.ciri.bam.csi
│   │   │   ├── GI1_N.ciri.log
│   │   │   ├── GI1_N.ciri.out
│   │   │   └── GI1_N.ciri.out.filtered
│   │   ├── CLEAR
│   │   │   ├── quant.txt
│   │   │   └── quant.txt.annotated
│   │   ├── DCC
│   │   │   ├── CircCoordinates
│   │   │   ├── CircRNACount
│   │   │   ├── CircSkipJunctions
│   │   │   ├── DCC-2023-12-20_1149.log
│   │   │   ├── GI1_N.dcc.counts_table.tsv
│   │   │   ├── GI1_N.dcc.counts_table.tsv.filtered
│   │   │   ├── LinearCount
│   │   │   ├── mate1.txt
│   │   │   ├── mate2.txt
│   │   │   └── samplesheet.txt
│   │   ├── find_circ
│   │   │   ├── GI1_N_anchors.fastq.gz
│   │   │   ├── GI1_N.unmapped.bam
│   │   │   └── GI1_N.unmapped.bam.csi
│   │   ├── GI1_N.circRNA_counts.txt.gz
│   │   ├── merge_per_sample.sh
│   │   ├── STAR1p
│   │   │   ├── GI1_N_p1.Chimeric.out.junction
│   │   │   ├── GI1_N_p1.Log.final.out
│   │   │   ├── GI1_N_p1.Log.out
│   │   │   ├── GI1_N_p1.Log.progress.out
│   │   │   ├── GI1_N_p1.SJ.out.tab
│   │   │   ├── mate1
│   │   │   │   ├── GI1_N_mate1.Chimeric.out.junction
│   │   │   │   ├── GI1_N_mate1.Log.final.out
│   │   │   │   ├── GI1_N_mate1.Log.out
│   │   │   │   ├── GI1_N_mate1.Log.progress.out
│   │   │   │   └── GI1_N_mate1.SJ.out.tab
│   │   │   └── mate2
│   │   │       ├── GI1_N_mate2.Chimeric.out.junction
│   │   │       ├── GI1_N_mate2.Chimeric.out.junction.fixed
│   │   │       ├── GI1_N_mate2.Log.final.out
│   │   │       ├── GI1_N_mate2.Log.out
│   │   │       ├── GI1_N_mate2.Log.progress.out
│   │   │       └── GI1_N_mate2.SJ.out.tab
│   │   ├── STAR2p
│   │   │   ├── GI1_N_p2.bam
│   │   │   ├── GI1_N_p2.bam.csi
│   │   │   ├── GI1_N_p2.chimeric.bam
│   │   │   ├── GI1_N_p2.chimeric.bam.csi
│   │   │   ├── GI1_N_p2.Chimeric.out.junction
│   │   │   ├── GI1_N_p2.Log.final.out
│   │   │   ├── GI1_N_p2.Log.out
│   │   │   ├── GI1_N_p2.Log.progress.out
│   │   │   ├── GI1_N_p2.non_chimeric.bam
│   │   │   ├── GI1_N_p2.non_chimeric.bam.csi
│   │   │   ├── GI1_N_p2.ReadsPerGene.out.tab
│   │   │   └── GI1_N_p2.SJ.out.tab
│   │   ├── STAR_circRNAFinder
│   │   │   ├── GI1_N.Chimeric.out.junction
│   │   │   ├── GI1_N.Chimeric.out.sam
│   │   │   ├── GI1_N.Log.final.out
│   │   │   ├── GI1_N.Log.out
│   │   │   ├── GI1_N.Log.progress.out
│   │   │   └── GI1_N.SJ.out.tab
│   │   └── trim
│   │       ├── GI1_N.R1.trim.fastq.gz
│   │       └── GI1_N.R2.trim.fastq.gz
│   ├── GI1_T
│   │   ├── alignmentstats.txt
│   │   ├── circExplorer
│   │   │   ├── GI1_T.back_spliced_junction.bed
│   │   │   ├── GI1_T.back_spliced_junction.strand_fixed.bed
│   │   │   ├── GI1_T.bam
│   │   │   ├── GI1_T.bam.bai
│   │   │   ├── GI1_T.BSJ.bam
│   │   │   ├── GI1_T.BSJ.bam.csi
│   │   │   ├── GI1_T.BSJ.bed.gz
│   │   │   ├── GI1_T.BSJ.bw
│   │   │   ├── GI1_T.BSJ.foundcounts.tsv
│   │   │   ├── GI1_T.BSJ.minus.bam
│   │   │   ├── GI1_T.BSJ.minus.bam.csi
│   │   │   ├── GI1_T.BSJ.minus.bw
│   │   │   ├── GI1_T.BSJ.plus.bam
│   │   │   ├── GI1_T.BSJ.plus.bam.csi
│   │   │   ├── GI1_T.BSJ.plus.bw
│   │   │   ├── GI1_T.circExplorer.annotation_counts.tsv
│   │   │   ├── GI1_T.circExplorer.counts_table.tsv
│   │   │   ├── GI1_T_circexplorer_parse.log
│   │   │   ├── GI1_T.circRNA_known.filter1.txt
│   │   │   ├── GI1_T.circularRNA_known.txt
│   │   │   ├── GI1_T.hg38.BSJ.bam
│   │   │   ├── GI1_T.hg38.BSJ.bam.csi
│   │   │   ├── GI1_T.hg38.BSJ.bw
│   │   │   ├── GI1_T.linear.bam
│   │   │   ├── GI1_T.linear.bam.csi
│   │   │   ├── GI1_T.linear_BSJ.bam
│   │   │   ├── GI1_T.linear_BSJ.bam.csi
│   │   │   ├── GI1_T.linear_BSJ.bw
│   │   │   ├── GI1_T.linear_BSJ.hg38.bam
│   │   │   ├── GI1_T.linear_BSJ.hg38.bw
│   │   │   ├── GI1_T.linear_BSJ.NC_009333.1.bam
│   │   │   ├── GI1_T.linear_BSJ.NC_009333.1.bw
│   │   │   ├── GI1_T.linear.BSJ.readids.gz
│   │   │   ├── GI1_T.linear.bw
│   │   │   ├── GI1_T.linear.hg38.bam
│   │   │   ├── GI1_T.linear.hg38.bw
│   │   │   ├── GI1_T.linear.NC_009333.1.bam
│   │   │   ├── GI1_T.linear.NC_009333.1.bw
│   │   │   ├── GI1_T.linear_spliced.counts.tsv
│   │   │   ├── GI1_T.NC_009333.1.BSJ.bam
│   │   │   ├── GI1_T.NC_009333.1.BSJ.bam.csi
│   │   │   ├── GI1_T.NC_009333.1.BSJ.bw
│   │   │   ├── GI1_T.readcounts.tsv
│   │   │   ├── GI1_T.rid2jid.tsv.gz
│   │   │   ├── GI1_T.spliced.bam
│   │   │   ├── GI1_T.spliced.bam.csi
│   │   │   ├── GI1_T.spliced_BSJ.bam
│   │   │   ├── GI1_T.spliced_BSJ.bam.csi
│   │   │   ├── GI1_T.spliced_BSJ.bw
│   │   │   ├── GI1_T.spliced_BSJ.hg38.bam
│   │   │   ├── GI1_T.spliced_BSJ.hg38.bw
│   │   │   ├── GI1_T.spliced_BSJ.NC_009333.1.bam
│   │   │   ├── GI1_T.spliced_BSJ.NC_009333.1.bw
│   │   │   ├── GI1_T.spliced.BSJ.readids.gz
│   │   │   ├── GI1_T.spliced.bw
│   │   │   ├── GI1_T.spliced.hg38.bam
│   │   │   ├── GI1_T.spliced.hg38.bw
│   │   │   ├── GI1_T.spliced.NC_009333.1.bam
│   │   │   ├── GI1_T.spliced.NC_009333.1.bw
│   │   │   ├── low_conf_GI1_T.circRNA_known.filter1.txt
│   │   │   └── low_conf_GI1_T.circRNA_known.txt
│   │   ├── circExplorer_BWA
│   │   │   ├── GI1_T.back_spliced_junction.bed
│   │   │   ├── GI1_T.back_spliced_junction.strand_fixed.bed
│   │   │   ├── GI1_T.circExplorer_bwa.annotation_counts.tsv
│   │   │   ├── GI1_T_circexplorer_parse.log
│   │   │   ├── GI1_T.circRNA_known.filter1.txt
│   │   │   ├── GI1_T.circularRNA_known.txt
│   │   │   ├── low_conf_GI1_T.circRNA_known.filter1.txt
│   │   │   └── low_conf_GI1_T.circRNA_known.txt
│   │   ├── circRNA_finder
│   │   │   ├── GI1_T.Chimeric.out.sorted.bam
│   │   │   ├── GI1_T.Chimeric.out.sorted.bam.bai
│   │   │   ├── GI1_T.circRNA_finder.counts_table.tsv.filtered
│   │   │   ├── GI1_T.filteredJunctions.bed
│   │   │   ├── GI1_T.s_filteredJunctions.bed
│   │   │   └── GI1_T.s_filteredJunctions_fw.bed
│   │   ├── ciri
│   │   │   ├── CIRIerror.log
│   │   │   ├── GI1_T.bwa.log
│   │   │   ├── GI1_T.ciri.bam
│   │   │   ├── GI1_T.ciri.bam.csi
│   │   │   ├── GI1_T.ciri.log
│   │   │   ├── GI1_T.ciri.out
│   │   │   └── GI1_T.ciri.out.filtered
│   │   ├── CLEAR
│   │   │   ├── quant.txt
│   │   │   └── quant.txt.annotated
│   │   ├── DCC
│   │   │   ├── CircCoordinates
│   │   │   ├── CircRNACount
│   │   │   ├── CircSkipJunctions
│   │   │   ├── DCC-2023-12-20_1206.log
│   │   │   ├── GI1_T.dcc.counts_table.tsv
│   │   │   ├── GI1_T.dcc.counts_table.tsv.filtered
│   │   │   ├── LinearCount
│   │   │   ├── mate1.txt
│   │   │   ├── mate2.txt
│   │   │   └── samplesheet.txt
│   │   ├── find_circ
│   │   │   ├── GI1_T_anchors.fastq.gz
│   │   │   ├── GI1_T.unmapped.bam
│   │   │   └── GI1_T.unmapped.bam.csi
│   │   ├── GI1_T.circRNA_counts.txt.gz
│   │   ├── merge_per_sample.sh
│   │   ├── STAR1p
│   │   │   ├── GI1_T_p1.Chimeric.out.junction
│   │   │   ├── GI1_T_p1.Chimeric.out.junction.circRNA
│   │   │   ├── GI1_T_p1.Chimeric.out.junction.circRNAmapped
│   │   │   ├── GI1_T_p1.Log.final.out
│   │   │   ├── GI1_T_p1.Log.out
│   │   │   ├── GI1_T_p1.Log.progress.out
│   │   │   ├── GI1_T_p1.SJ.out.tab
│   │   │   ├── mate1
│   │   │   │   └── GI1_T_mate1.Chimeric.out.junction
│   │   │   └── mate2
│   │   │       └── GI1_T_mate2.Chimeric.out.junction
│   │   ├── STAR2p
│   │   │   ├── GI1_T_p2.bam
│   │   │   ├── GI1_T_p2.bam.csi
│   │   │   ├── GI1_T_p2.chimeric.bam
│   │   │   ├── GI1_T_p2.chimeric.bam.csi
│   │   │   ├── GI1_T_p2.Chimeric.out.junction
│   │   │   ├── GI1_T_p2.Log.final.out
│   │   │   ├── GI1_T_p2.Log.out
│   │   │   ├── GI1_T_p2.Log.progress.out
│   │   │   ├── GI1_T_p2.non_chimeric.bam
│   │   │   ├── GI1_T_p2.non_chimeric.bam.csi
│   │   │   ├── GI1_T_p2.ReadsPerGene.out.tab
│   │   │   └── GI1_T_p2.SJ.out.tab
│   │   ├── STAR_circRNAFinder
│   │   │   ├── GI1_T.Chimeric.out.junction
│   │   │   ├── GI1_T.Chimeric.out.sam
│   │   │   ├── GI1_T.Log.final.out
│   │   │   ├── GI1_T.Log.out
│   │   │   ├── GI1_T.Log.progress.out
│   │   │   └── GI1_T.SJ.out.tab
│   │   └── trim
│   │       ├── GI1_T.R1.trim.fastq.gz
│   │       └── GI1_T.R2.trim.fastq.gz
│   ├── HQ_BSJ_bams
│   │   ├── GI1_N.hg38.HQ_only.BSJ.bam
│   │   ├── GI1_N.hg38.HQ_only.BSJ.bam.bai
│   │   ├── GI1_N.HQ_only.BSJ.bam
│   │   ├── GI1_N.HQ_only.BSJ.bam.bai
│   │   ├── GI1_N.NC_009333.1.HQ_only.BSJ.bam
│   │   ├── GI1_N.NC_009333.1.HQ_only.BSJ.bam.bai
│   │   ├── GI1_T.hg38.HQ_only.BSJ.bam
│   │   ├── GI1_T.hg38.HQ_only.BSJ.bam.bai
│   │   ├── GI1_T.HQ_only.BSJ.bam
│   │   ├── GI1_T.HQ_only.BSJ.bam.bai
│   │   ├── GI1_T.NC_009333.1.HQ_only.BSJ.bam
│   │   └── GI1_T.NC_009333.1.HQ_only.BSJ.bam.bai
│   ├── pass1.out.tab
│   ├── revstranded_STAR_GeneCounts.tsv
│   ├── stranded_STAR_GeneCounts.tsv
│   └── unstranded_STAR_GeneCounts.tsv
├── runinfo.yaml
├── run_jobby_on_snakemake_log
├── runslurm_snakemake_report.html
├── samples.tsv
├── snakemake.log
├── snakemake.stats
├── stats
│   ├── dryrun.231222102114.log
│   ├── runinfo.modtime.yaml
│   └── snakemake.20231222104214.log
├── submit_script.sbatch
└── tools.txt
</code></pre>
<h4 id="error-tracking">Error tracking</h4>
<p>Using the following command to find FAILED jobs:</p>
<pre><code class="language-bash">grep FAIL jobinfo.short
</code></pre>
<p>The above command also gives <code>.err</code> and <code>.out</code> log files which can give further insights on reasons for failure and changes required to be made for a successful run.</p>
<h4 id="expected-output">Expected output:</h4>
<p>The main output file is <code>results/circRNA_master_counts.tsv.gz</code>. Here are the top 3 tiles from an example output:</p>
<table>
<thead>
<tr>
<th>Column_number</th>
<th>Column_title</th>
<th>Example_1</th>
<th>Example_2</th>
<th>Example_3</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>chrom</td>
<td>GL000220.1</td>
<td>GL000220.1</td>
<td>GL000220.1</td>
</tr>
<tr>
<td>2</td>
<td>start</td>
<td>107635</td>
<td>112482</td>
<td>118578</td>
</tr>
<tr>
<td>3</td>
<td>end</td>
<td>151634</td>
<td>156427</td>
<td>118759</td>
</tr>
<tr>
<td>4</td>
<td>circExplorer_strand</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>5</td>
<td>circExplorer_bwa_strand</td>
<td>.</td>
<td>.</td>
<td>.</td>
</tr>
<tr>
<td>6</td>
<td>ciri_strand</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>7</td>
<td>dcc_strand</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>8</td>
<td>circrnafinder_strand</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>9</td>
<td>flanking_sites_+</td>
<td>CC##GC</td>
<td>GC##CC</td>
<td>CC##GC</td>
</tr>
<tr>
<td>10</td>
<td>flanking_sites_-</td>
<td>GG##GC</td>
<td>GC##GG</td>
<td>GG##GC</td>
</tr>
<tr>
<td>11</td>
<td>sample_name</td>
<td>GI1_N</td>
<td>GI1_N</td>
<td>GI1_N</td>
</tr>
<tr>
<td>12</td>
<td>ntools</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>13</td>
<td>HQ</td>
<td>N</td>
<td>N</td>
<td>N</td>
</tr>
<tr>
<td>14</td>
<td>circExplorer_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>15</td>
<td>circExplorer_found_BSJcounts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>16</td>
<td>circExplorer_found_linear_BSJ_+_counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>17</td>
<td>circExplorer_found_linear_spliced_BSJ_+_counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>18</td>
<td>circExplorer_found_linear_BSJ_-_counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>19</td>
<td>circExplorer_found_linear_spliced_BSJ_-_counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>20</td>
<td>circExplorer_found_linear_BSJ_._counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>21</td>
<td>circExplorer_found_linear_spliced_BSJ_._counts</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>22</td>
<td>ciri_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>23</td>
<td>ciri_linear_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>24</td>
<td>circExplorer_bwa_read_count</td>
<td>3</td>
<td>7</td>
<td>3</td>
</tr>
<tr>
<td>25</td>
<td>dcc_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>26</td>
<td>dcc_linear_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>27</td>
<td>circrnafinder_read_count</td>
<td>-1</td>
<td>-1</td>
<td>-1</td>
</tr>
<tr>
<td>28</td>
<td>hqcounts</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>29</td>
<td>nonhqcounts</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>30</td>
<td>circExplorer_annotation</td>
<td>Unknown</td>
<td>Unknown</td>
<td>Unknown</td>
</tr>
<tr>
<td>31</td>
<td>ciri_annotation</td>
<td>Unknown</td>
<td>Unknown</td>
<td>Unknown</td>
</tr>
<tr>
<td>32</td>
<td>circExplorer_bwa_annotation</td>
<td>novel</td>
<td>novel</td>
<td>novel</td>
</tr>
<tr>
<td>33</td>
<td>dcc_gene</td>
<td>Unknown</td>
<td>Unknown</td>
<td>Unknown</td>
</tr>
<tr>
<td>34</td>
<td>dcc_junction_type</td>
<td>Unknown</td>
<td>Unknown</td>
<td>Unknown</td>
</tr>
<tr>
<td>35</td>
<td>dcc_annotation</td>
<td>Unknown</td>
<td>Unknown</td>
<td>Unknown</td>
</tr>
</tbody>
</table>
<p>Expected output from the sample data is stored under <code>.tests/expected_output</code>.</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2023 CCBR
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="http://bioinformatics.cancer.gov" target="_blank" rel="noopener" title="bioinformatics.cancer.gov" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/CCBR" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://hub.docker.com/orgs/nciccbr/repositories" target="_blank" rel="noopener" title="hub.docker.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.51198bba.min.js"></script>
      
    
  </body>
</html>